### PR TITLE
feature: Apply commit label on resources created/updated when '--commitlabel' is passed

### LINKED ISSUE #2208: Append commit as label when apply spec
**Is your feature request related to a problem? Please describe.**
When using the `fission spec apply`, it is hard to know which version we are deploying now. Furthermore, we cannot easily know which version has been deployed to the cluster before.

**Describe the solution you'd like**
I would like to have a label on the components included in the specs. Each time, when I use `fission spec apply`, this label can update automatically so that I can know which version is deploying now. This label represents git commit so that we can identify it easily.

### PR DESCRIPTION
**Pull request by praddy26**
## Description
When `--commitlabel` is passed to the `fission spec apply` command, the resources created will have `commit`
 label with value set depending upon the status of the spec file in the git repository. If the `specdir` is not part of git repository then despite passing the `--commitlabel` flag this label will not be applied on the resources created/updated.

The value of the `commit` label for different status of the file is as follows:
| Git File Status | Label Value |
| :--- | :---: |
| New untracked file | `untracked` |
| New staged file | `staged` |
| Tracked file with changes in worktree | `<commitID>-unstaged` |
| Tracked file with changes staged | `<commitID>-staged` |
| Tracked file with clean commit | `<commitID>` |

The value of `<commitID>` will be the short commit version of the commit the HEAD will be pointing to.

**P.S**: This PR also contains the changes to update resource if there is a difference in the `labels` and `annotations` of the reources in the spec file and the existing object. Prior to this only changes under `spec` were considered for updating the resource.

## Which issue(s) this PR fixes:

Fixes #2208 

## Testing

Before changes
------------------------
4.59s user 0.71s system 92% cpu 5.749
3.83s user 0.70s system 85% cpu 5.281
4.14s user 0.71s system 88% cpu 5.492

After changes
----------------------
(gitrepo)
4.38s user 0.73s system 90% cpu 5.644
3.93s user 0.70s system 87% cpu 5.316  

(non-gitrepo)
3.83s user 0.67s system 84% cpu 5.302  
3.88s user 0.68s system 84% cpu 5.410

## Checklist:

- [x] I ran tests as well as code linting locally to verify my changes. 
- [x] I have done manual verification of my changes, changes working as expected.
- [ ] I have added new tests to cover my changes.
- [x] My changes follow contributing guidelines of Fission.
- [x] I have signed all of my commits.

### CODE REVIEW

File: pkg/fission-cli/cmd/spec/validate.go [hunk 1]
```diff
@@ -225,6 +226,55 @@ func ReadSpecs(specDir, specIgnore string) (*FissionResources, error) {
...
 	}
 
+	var isGitRepo bool
+	var commitID string
+	var repo *git.Repository
+	var status git.Status
+	var gitRepoRootPath string
+
+	// check if applyCommitLabel flag is true
+	if applyCommitLabel {
```
[the remark this answers is not in the document]
Comment (sanketsudake) [reply pkg/fission-cli/cmd/spec/validate.go:236]: Also there is `warnIfDirtyWorkTree` function in spec/apply.go . Please check if we can use it here.

File: pkg/fission-cli/cmd/spec/validate.go [hunk 2]
```diff
@@ -281,3 +339,59 @@ func ReadSpecs(specDir, specIgnore string) (*FissionResources, error) {
...
+		}
+	}
+	return -1
+}
+
+// getCommitLabel returns the value of the 'commit' label
+func getCommitLabel(path, gitRepoRootPath, commitID string, status *git.Status, repo *git.Repository) string {
+
+	gitRepoDirName := filepath.Base(gitRepoRootPath)
+	splitPathList := strings.Split(path, string(os.PathSeparator))
```
Comment (sanketsudake) [inline pkg/fission-cli/cmd/spec/validate.go:356]: Please see if we can do this part without splitting the string. Maybe use regex. index function seems unnecessary to me. 
Also, we are not handling the -1 return case.

**Review by sanketsudake — CHANGES_REQUESTED** [review]

File: pkg/fission-cli/cmd/spec/validate.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (praddy26) [reply pkg/fission-cli/cmd/spec/validate.go:236]: > 

In the `if applyLabel` section we are checking if it's git repository and extracting commitID.
But in the `if applyCommitLabel && isGitRepo` which is within the walkdirectory method we are investigating the status of each file and accordingly creating the label by calling `getCommitLabel` method. Hence we can't move this under `if applyLabel`

File: pkg/fission-cli/cmd/spec/validate.go [hunk 3]
```diff
@@ -225,6 +226,55 @@ func ReadSpecs(specDir, specIgnore string) (*FissionResources, error) {
...
+	var status git.Status
+	var gitRepoRootPath string
+
+	// check if applyCommitLabel flag is true
+	if applyCommitLabel {
+
+		// check if spec directory is tracked by git repo
+		repo, err = git.PlainOpenWithOptions(specDir, &git.PlainOpenOptions{DetectDotGit: true})
+		if err != nil {
+			goto asbolutePath
```
[the remark this answers is not in the document]
Comment (praddy26) [reply pkg/fission-cli/cmd/spec/validate.go:241]: Corrected

File: pkg/fission-cli/cmd/spec/apply.go [hunk 4]
```diff
@@ -1184,3 +1184,23 @@ func applyMessageQueueTriggers(fclient client.Interface, fr *FissionResources, d
 
 	return metadataMap, &ras, nil
 }
+
+func isObjectMetaEqual(existingObj, newObj metav1.ObjectMeta) bool {
+
+	x := make(map[string]string)
```
[the remark this answers is not in the document]
Comment (praddy26) [reply pkg/fission-cli/cmd/spec/apply.go:1190]: Corrected

File: pkg/fission-cli/cmd/spec/apply.go [hunk 5]
```diff
@@ -1184,3 +1184,23 @@ func applyMessageQueueTriggers(fclient client.Interface, fr *FissionResources, d
...
+		}
+	}
+
+	for newObjKey, newObjVal := range newObj.Annotations {
+		if newObjKey != FISSION_DEPLOYMENT_NAME_KEY && newObjKey != FISSION_DEPLOYMENT_UID_KEY {
+			y[newObjKey] = newObjVal
+		}
+	}
+
+	return reflect.DeepEqual(existingObj.Labels, newObj.Labels) && reflect.DeepEqual(x, y)
```
[the remark this answers is not in the document]
Comment (praddy26) [reply pkg/fission-cli/cmd/spec/apply.go:1205]: Corrected

File: pkg/fission-cli/cmd/spec/validate.go (hunk 2 above)
Comment (praddy26) [reply pkg/fission-cli/cmd/spec/validate.go]: Here I have the absolute path of the files in the specdir and I have the absolute path of the git repository, where spec directory can be the root directory of the git repo or it might be nested inside the git repo tree. So here I am removing the absolute path of the git repo from the absolute specdir file path to get the relative file apth inside the git repository. So this has to exist and the -1 case has been handled. The index function has been removed.

File: charts/fission-all/values.yaml [hunk 6]
```diff
@@ -386,7 +386,7 @@ persistence:
 ## Analytics let us count how many people installed fission. Set to
 ## false to disable analytics.
 ##
-analytics: true
+analytics: false
```
Comment (sanketsudake) [inline charts/fission-all/values.yaml:389]: Please revert this change.

File: pkg/fission-cli/cmd/spec/validate.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (praddy26) [reply pkg/fission-cli/cmd/spec/validate.go:231]: corrected

File: pkg/utils/gitrepo/gitrepo.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	IsGitRepo       bool
+	repo            *git.Repository
+	status          git.Status
+	gitRepoRootPath string // absolute path of the root of the git repository
+	commitID        string // commit ID the HEAD of the repository is poitning to
+}
+
+// CheckIsGitRepo checks if the given directory path is git repo
+// and updates the repo, status, gitRepoRootPath and commitID of the Gitrepo
+func (g *Gitrepo) CheckIsGitRepo(dirPath string) {
```
Comment (sanketsudake) [inline pkg/utils/gitrepo/gitrepo.go:38]:
```
 func NewGitRepo() (*GitRepo, error) {
   gr = &GitRepo{}
   gr.Setup()
}


func (g *Gitrepo) Setup(dirPath string) error
func (g *Gitrepo) IsGitRepo() bool {
    return setupDone && isGitRepo
}
func (g *Gitrepo) GetFileCommitLabel(filePath string) (string, error)

type Gitrepo struct {
       setupDone       bool
	isGitRepo       bool
	repo            *git.Repository
	status          git.Status
	gitRepoRootPath string // absolute path of the root of the git repository
	commitID        string // commit ID the HEAD of the repository is poitning to
}
```

Comment (praddy26) [reply pkg/utils/gitrepo/gitrepo.go]: Fixed

File: pkg/fission-cli/cmd/spec/validate.go [hunk 8]
```diff
@@ -239,16 +255,24 @@ func ReadSpecs(specDir, specIgnore string) (*FissionResources, error) {
...
 
+		// check if file matches any path in .specignore file
 		if ignoreParser.MatchesPath(path) {
 			return nil
 		}
 
+		var fileCommitLabelVal string
+		// check if applyCommitLabel is true and specdir is tracked by git repo
+		if applyCommitLabel {
+			fileCommitLabelVal, _ = gr.GetFileCommitLabel(path)
```
[the remark this answers is not in the document]
Comment (praddy26) [reply pkg/fission-cli/cmd/spec/validate.go:266]: Actually this function is called twice once for validation and other for apply. In that case, it would print the warining twice.

[the remark this answers is not in the document]
Comment (praddy26) [reply pkg/fission-cli/cmd/spec/validate.go:266]: Also in the apply function there is a call to warn if the specs is not part of git tree or if it's a dirty work tree, calling warnIfDirtyWorkTree function.

**Review by sanketsudake — APPROVED** [review]
